overlord/state/notices_test.go

```diff
@@ -0,0 +1,533 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package state_test
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+type noticesSuite struct{}
+
+var _ = Suite(&noticesSuite{})
+
+func (s *noticesSuite) TestMarshal(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	start := time.Now()
+	addNotice(c, st, state.ChangeUpdateNotice, "123", nil)
+	time.Sleep(time.Microsecond) // ensure there's time between the occurrences
+	addNotice(c, st, state.ChangeUpdateNotice, "123", &state.AddNoticeOptions{
+		Data: map[string]string{"k": "v"},
+	})
+
+	notices := st.Notices(nil)
+	c.Assert(notices, HasLen, 1)
+
+	// Convert it to a map so we're not testing the JSON string directly
+	// (order of fields doesn't matter).
+	n := noticeToMap(c, notices[0])
+
+	firstOccurred, err := time.Parse(time.RFC3339, n["first-occurred"].(string))
+	c.Assert(err, IsNil)
+	c.Assert(!firstOccurred.Before(start), Equals, true) // firstOccurred >= start
+	lastOccurred, err := time.Parse(time.RFC3339, n["last-occurred"].(string))
+	c.Assert(err, IsNil)
+	c.Assert(lastOccurred.After(firstOccurred), Equals, true) // lastOccurred > firstOccurred
+	lastRepeated, err := time.Parse(time.RFC3339, n["last-repeated"].(string))
+	c.Assert(err, IsNil)
+	c.Assert(lastRepeated.After(firstOccurred), Equals, true) // lastRepeated > firstOccurred
+
+	delete(n, "first-occurred")
+	delete(n, "last-occurred")
+	delete(n, "last-repeated")
+	c.Assert(n, DeepEquals, map[string]any{
+		"id":           "1",
+		"type":         "change-update",
+		"key":          "123",
+		"occurrences":  2.0,
+		"last-data":    map[string]any{"k": "v"},
+		"expire-after": "168h0m0s",
+	})
+}
+
+func (s *noticesSuite) TestUnmarshal(c *C) {
+	noticeJSON := []byte(`{
+		"id": "1",
+		"type": "change-update",
+		"key": "123",
+		"first-occurred": "2023-09-01T05:23:01Z",
+		"last-occurred": "2023-09-01T07:23:02Z",
+		"last-repeated": "2023-09-01T06:23:03.123456789Z",
+		"occurrences": 2,
+		"last-data": {"k": "v"},
+		"repeat-after": "60m",
+		"expire-after": "168h0m0s"
+	}`)
+	var notice *state.Notice
+	err := json.Unmarshal(noticeJSON, &notice)
+	c.Assert(err, IsNil)
+
+	// The Notice fields aren't exported, so we need to marshal it into JSON
+	// and then unmarshal it into a map to test.
+	n := noticeToMap(c, notice)
+	c.Assert(n, DeepEquals, map[string]any{
+		"id":             "1",
+		"type":           "change-update",
+		"key":            "123",
+		"first-occurred": "2023-09-01T05:23:01Z",
+		"last-occurred":  "2023-09-01T07:23:02Z",
+		"last-repeated":  "2023-09-01T06:23:03.123456789Z",
+		"occurrences":    2.0,
```

Comment:
Integer values in notices (here, `"occurrences"`) are marshalled as floats. As a result, a notice -> marshal -> unmarshal fails because the float is not an int. I think there should be logic in the `Marshal`/`Unmarshal` functions to be able to handle this, as it seems JSON marshals all numbers as floats. Some options I see:
1. Use `float` in `jsonNotice` in place of integer types, and then convert from float to integer when converting the `jsonNotice` to a `Notice`.
2. Store integer types as strings in base-10, so there is no mangling of negatives or floating-point imprecision

I would favor the latter if JSON will be happy about working with strings which represent integers. If not, I think something like the former is needed so that Notice -> marshal -> unmarshal -> Notice works. I checked, and there should not be any floating point error on floats between -1 and UINT32_MAX.

Reply:
This is not actually a problem. Marshalling/unmarshalling works just fine, as expected.

I changed the `"occurrences"` field to a float *within the json string*, which caused unmarshalling to fail, but that is expected as it is in the string. So there's no real problem.